File: Frontend/src/config/print-template-catalog.ts
/** Print template codes — mirror Backend print_template_catalog.py */

export interface PrintTemplateMeta {
  code: string;
  label: string;
  group: string;
  supportsTwoCopies?: boolean;
  printModes?: string[];
}

export const PRINT_TEMPLATE_CATALOG: PrintTemplateMeta[] = [
  { code: "si", label: "Sales Invoice (SI)", group: "Sales Printing" },
  { code: "sc", label: "Sales Credit (SC)", group: "Sales Printing" },
  { code: "so", label: "Sales Order (SO)", group: "Sales Printing" },
  { code: "sr", label: "Sales Receipt (SR)", group: "Sales Printing", supportsTwoCopies: true },
  { code: "pdcr", label: "Post Dated Cheque Received (PDCR)", group: "Sales Printing" },
  { code: "gdnso", label: "Delivery Note from SO (GDNSO)", group: "Sales Printing" },
  { code: "gdnsi", label: "Delivery Note from SI (GDNSI)", group: "Sales Printing" },
  { code: "cus", label: "Customer (CUS)", group: "Sales Printing" },
  { code: "vi", label: "Supplier Bill (VI)", group: "Purchases Printing" },
  { code: "vc", label: "Supplier Credit (VC)", group: "Purchases Printing" },
  { code: "po", label: "Purchase Order (PO)", group: "Purchases Printing" },
  { code: "vp", label: "Bill Payment (VP)", group: "Purchases Printing", supportsTwoCopies: true },
  { code: "pdci", label: "Post Dated Cheque Issued (PDCI)", group: "Purchases Printing" },
  { code: "grnpo", label: "GRN from PO (GRNPO)", group: "Purchases Printing" },
  { code: "grnvi", label: "GRN from VI (GRNVI)", group: "Purchases Printing" },
  { code: "journal", label: "Journal", group: "Journal / Bank / Other / Project Printing" },
  {
    code: "bank",
    label: "Bank voucher",
    group: "Journal / Bank / Other / Project Printing",
    printModes: ["voucher", "journal"],
  },
  { code: "assembly", label: "Assembly", group: "Journal / Bank / Other / Project Printing" },
  { code: "stock-adjustment", label: "Stock Adjustment", group: "Journal / Bank / Other / Project Printing" },
  { code: "stock-transfer", label: "Stock Transfer", group: "Journal / Bank / Other / Project Printing" },
  { code: "user-log", label: "User Log", group: "Journal / Bank / Other / Project Printing" },
  { code: "project", label: "Project", group: "Journal / Bank / Other / Project Printing" },
];

export const PRINT_TEMPLATE_BY_CODE = Object.fromEntries(
  PRINT_TEMPLATE_CATALOG.map((t) => [t.code, t]),
) as Record<string, PrintTemplateMeta>;

export function printTemplateMeta(code: string): PrintTemplateMeta | undefined {
  return PRINT_TEMPLATE_BY_CODE[code];
}

File: Backend/src/app/constants/print_template_catalog.py
"""Print template codes — FastAccounts Settings → Printing (catalog §12.1 / §12.8)."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class PrintTemplateDef:
    code: str
    label: str
    group: str
    supports_two_copies: bool = False
    print_modes: tuple[str, ...] = ("document",)


PRINT_TEMPLATE_CATALOG: tuple[PrintTemplateDef, ...] = (
    # Sales
    PrintTemplateDef("si", "Sales Invoice (SI)", "Sales Printing"),
    PrintTemplateDef("sc", "Sales Credit (SC)", "Sales Printing"),
    PrintTemplateDef("so", "Sales Order (SO)", "Sales Printing"),
    PrintTemplateDef("sr", "Sales Receipt (SR)", "Sales Printing", supports_two_copies=True),
    PrintTemplateDef("pdcr", "Post Dated Cheque Received (PDCR)", "Sales Printing"),
    PrintTemplateDef("gdnso", "Delivery Note from SO (GDNSO)", "Sales Printing"),
    PrintTemplateDef("gdnsi", "Delivery Note from SI (GDNSI)", "Sales Printing"),
    PrintTemplateDef("cus", "Customer (CUS)", "Sales Printing"),
    # Purchases
    PrintTemplateDef("vi", "Supplier Bill (VI)", "Purchases Printing"),
    PrintTemplateDef("vc", "Supplier Credit (VC)", "Purchases Printing"),
    PrintTemplateDef("po", "Purchase Order (PO)", "Purchases Printing"),
    PrintTemplateDef("vp", "Bill Payment (VP)", "Purchases Printing", supports_two_copies=True),
    PrintTemplateDef("pdci", "Post Dated Cheque Issued (PDCI)", "Purchases Printing"),
    PrintTemplateDef("grnpo", "GRN from PO (GRNPO)", "Purchases Printing"),
    PrintTemplateDef("grnvi", "GRN from VI (GRNVI)", "Purchases Printing"),
    # Journal / Bank / Other / Project
    PrintTemplateDef("journal", "Journal", "Journal / Bank / Other / Project Printing"),
    PrintTemplateDef(
        "bank",
        "Bank voucher",
        "Journal / Bank / Other / Project Printing",
        print_modes=("voucher", "journal"),
    ),
    PrintTemplateDef("assembly", "Assembly", "Journal / Bank / Other / Project Printing"),
    PrintTemplateDef("stock-adjustment", "Stock Adjustment", "Journal / Bank / Other / Project Printing"),
    PrintTemplateDef("stock-transfer", "Stock Transfer", "Journal / Bank / Other / Project Printing"),
    PrintTemplateDef("user-log", "User Log", "Journal / Bank / Other / Project Printing"),
    PrintTemplateDef("project", "Project", "Journal / Bank / Other / Project Printing"),
)

PRINT_TEMPLATE_BY_CODE: dict[str, PrintTemplateDef] = {t.code: t for t in PRINT_TEMPLATE_CATALOG}


def default_print_template(code: str) -> dict[str, object]:
    """Default layout flags for a document print template."""

    meta = PRINT_TEMPLATE_BY_CODE.get(code)
    out: dict[str, object] = {
        "code": code,
        "title": meta.label if meta else code.upper(),
        "showLogo": True,
        "showBusinessBlock": True,
        "headerNote": "",
        "footerNote": "",
        "paperSize": "A4",
        "showTaxColumns": True,
        "printMode": "document",
    }
    if meta and meta.supports_two_copies:
        out["twoCopies"] = False
    if meta and len(meta.print_modes) > 1:
        out["printMode"] = meta.print_modes[0]
    return out
